rusk/src/lib/chain/rusk.rs

```diff
@@ -540,7 +540,11 @@ fn contract_deploy(
         receipt.data = Err(Panic("failed bytecode hash check".into()))
     } else {
         let result = session.deploy_raw(
-            None,
+            Some(gen_contract_id(
```

Comment:
In general I agree with your remarks. I am not sure what do you mean by "the nonce is checked against the owner" but I'd consider generating contract id just randomly, and give up on the idea of contract id as hash at all. This would mean we would not have vanity ids though, but this is fine IMO. Let's discuss this.

Reply:
I guess we can live without vanity IDs :)

What I meant by "the nonce is checked against the owner" is that we enforce different nonces for each owner (I'm just guessing here, since the nonce check seems not to be done here). I'm assuming it's the owner because we can't use the sender if the deploy transaction is made with Phoenix, as far as I understand.
So, if the nonce is unique only in relation to the owner, two different owners could deploy the same bytecode with the same nonce value.

My suggestion is to tie the ID to the hash of the owner id plus the nonce. If we enforce the nonce to change each time the owner deploys a contract, this ensures all contracts have unique IDs.

I'm open to discussion of course

Reply:
I see, I think it is a very good idea to include owner in the scheme. Currently, owner is not used, and nonce is introduced only for the purpose of vanity ids. If we include owner in hash, we could still have vanity id and owner uniqueness. I don't see the need for nonce uniqueness between owners, tbh.

Reply:
Having a nonce allows an owner to deploy different contracts with different IDs (without the need of specifying additional data). Without it you would depend on the metadata to obtain a different contract IDs.
As an extra, using a unique nonce per owner, you don't need to check if a contract ID is already used.

Reply:
I added owner to contract id generation method.